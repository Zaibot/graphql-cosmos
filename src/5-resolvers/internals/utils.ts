import { FeedResponse } from '@azure/cosmos'
import { GraphQLCosmosPageOutput } from '../../4-resolver-builder/3-typedefs-transformer'
import { Lazy } from '../../typescript'
import { SourceDescriptor } from '../x-descriptors'

export const emptyPageResponse: GraphQLCosmosPageOutput<any> = {
  cursor: null,
  nextCursor: null,
  page: [],
  total: 0,
}

export function graphqlCosmosPageResponse<T extends { id: string }>(
  cursor: string | null,
  query: Lazy<Promise<FeedResponse<T>>>,
  count: Lazy<Promise<FeedResponse<number>>>,
  transformer: (item: T) => SourceDescriptor.Embedded<T>
): GraphQLCosmosPageOutput<T> {
  return {
    cursor: cursor,
    nextCursor: () => query().then((x) => x.continuationToken),
    page: () => query().then((x) => (x.resources ?? []).map(transformer)),
    total: () => count().then((x) => x.resources[0]),
  }
}
