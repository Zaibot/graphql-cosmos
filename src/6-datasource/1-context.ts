import { GraphQLCosmosDataSource } from './2-datasource'

export interface GraphQLCosmosConceptContext<TContext = unknown> {
  dataSources: {
    graphqlCosmos: GraphQLCosmosDataSource<TContext>
  }
}
