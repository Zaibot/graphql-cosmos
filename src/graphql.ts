import { GraphQLInputFieldConfig, GraphQLSchema, GraphQLInputType, GraphQLInputObjectType, GraphQLField } from 'graphql';

export const createOrGetWhereType = (name: string, properties: Record<string, GraphQLInputFieldConfig>, schema: GraphQLSchema) => {
    if (schema.getTypeMap()[name]) {
        return schema.getTypeMap()[name] as GraphQLInputType;
    }

    const filterType = new GraphQLInputObjectType({
        name,
        description: undefined,
        extensions: undefined,
        fields: properties,
        astNode: undefined,
        extensionASTNodes: undefined,
    });

    // HACK
    schema.getTypeMap()[name] = filterType;

    return filterType;
};

export const addFieldArgument = (field: GraphQLField<any, any, any>, name: string, type: GraphQLInputType) => {
    field.args.push({
        name,
        type,
        astNode: undefined,
        extensions: undefined,
        description: undefined,
        defaultValue: undefined,
    });
};
