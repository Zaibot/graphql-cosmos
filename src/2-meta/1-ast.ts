import {
  DefinitionNode,
  DocumentNode,
  FieldDefinitionNode,
  ListTypeNode,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  ObjectTypeExtensionNode,
  TypeNode,
} from 'graphql'
import { isListTypeNode } from 'graphql-tools'
import { Directives } from '../1-graphql/2-ast'
import { isWhereOp, WhereOps } from '../6-datasource/x-where'

export interface GraphQLCosmosSchemaFromAst {
  types: GraphQLCosmosTypeFromAst[]
}

export interface GraphQLCosmosTypeFromAst {
  external: boolean
  typename: string
  database: string | null
  container: string | null
  fields: GraphQLCosmosFieldFromAst[]
}

export interface GraphQLCosmosFieldFromAst {
  external: boolean
  fieldname: string
  returnTypename: string
  returnMany: boolean
  container: string | null
  database: string | null
  ours: string | null
  theirs: string | null
  whereOps: WhereOps[] | null
  whereOurs: string | null
  sortable: boolean | null
  sortOurs: string | null
  pagination: string | null
}

export function getGraphQLCosmosSchemaFromGraphQL(objectType: DocumentNode): GraphQLCosmosSchemaFromAst {
  const internalTypes = (objectType.definitions ?? []).filter(isObjectTypeDefinition).map(getGraphQLCosmosTypeFromGraphQL)
  const externalTypes = (objectType.definitions ?? []).filter(isObjectTypeExtension).map(getGraphQLCosmosTypeFromGraphQL)

  return {
    types: [...internalTypes, ...externalTypes],
  }
}

export function getGraphQLCosmosTypeFromGraphQL(objectType: ObjectTypeDefinitionNode | ObjectTypeExtensionNode): GraphQLCosmosTypeFromAst {
  const typename = objectType.name.value
  const fields = (objectType.fields ?? []).map(getGraphQLCosmosFieldFromGraphQL)
  const database = Directives.cosmosDatabaseDirective(objectType.directives ?? []) ?? null
  const container = Directives.cosmosContainerDirective(objectType.directives ?? []) ?? null

  return {
    external: objectType.kind === `ObjectTypeExtension`,
    typename: typename,
    database,
    container,
    fields,
  }
}

export function getGraphQLCosmosFieldFromGraphQL(field: FieldDefinitionNode): GraphQLCosmosFieldFromAst {
  const returnTypename = astGetNamedType(field.type).name.value
  const returnMany = isListTypeNode(astGetNullable(field.type))
  const database = Directives.cosmosDatabaseDirective(field.directives ?? []) ?? null
  const container = Directives.cosmosContainerDirective(field.directives ?? []) ?? null
  const pagination = Directives.cosmosPaginationDirective(field.directives ?? []) ?? null
  const fieldname = field.name.value
  const ours = Directives.cosmosOursDirective(field.directives ?? []) ?? null
  const theirs = Directives.cosmosTheirsDirective(field.directives ?? []) ?? null
  const whereOps = Directives.whereOpDirective(field.directives ?? []).filter(isWhereOp) ?? null
  const whereOurs = Directives.whereOursDirective(field.directives ?? []) ?? null
  const sortable = Directives.sortDirective(field.directives ?? []) ?? null
  const sortOurs = Directives.sortOursDirective(field.directives ?? []) ?? null

  return {
    external: false,
    returnTypename,
    returnMany,
    database,
    container,
    fieldname,
    ours,
    theirs,
    whereOps,
    whereOurs,
    sortable,
    sortOurs,
    pagination,
  }
}

function astGetNamedType(type: TypeNode): NamedTypeNode {
  while (type.kind === `ListType` || type.kind === `NonNullType`) {
    type = type.type
  }
  return type
}

function isObjectTypeDefinition(n: DefinitionNode): n is ObjectTypeDefinitionNode {
  return n.kind === `ObjectTypeDefinition`
}

function isObjectTypeExtension(n: DefinitionNode): n is ObjectTypeExtensionNode {
  return n.kind === `ObjectTypeExtension`
}

function astGetNullable(type: TypeNode): ListTypeNode | NamedTypeNode {
  while (type.kind === `NonNullType`) {
    type = type.type
  }
  return type
}
