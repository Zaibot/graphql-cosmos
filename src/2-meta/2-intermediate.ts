import { WhereOps } from '../6-datasource/x-where'
import { GraphQLCosmosInitError } from '../x-error/init-error'
import { GraphQLCosmosFieldFromAst, GraphQLCosmosSchemaFromAst, GraphQLCosmosTypeFromAst } from './1-ast'

export interface MetaSchema {
  types: MetaType[]
}

export interface MetaType {
  cosmos: boolean
  external: boolean
  typename: string
  database: string | null
  container: string | null
  fields: MetaField[]
  filterable: boolean | null
  sortable: boolean | null
}

export interface MetaField {
  cosmos: boolean
  fieldname: string
  returnMany: boolean
  returnTypename: string
  container: string | null
  database: string | null
  ours: string | null
  theirs: string | null
  pagination: boolean
  whereOps: WhereOps[] | null
  whereOurs: string | null
  sortable: boolean | null
  sortOurs: string | null
  // kind: MetaFieldKind
}

export type MetaFieldKind =
  | 'one-root'
  | 'many-root'
  | 'one-ours'
  | 'many-ours'
  | 'one-theirs'
  | 'many-theirs'
  | 'embedded'

export function getMetaSchema(schema: GraphQLCosmosSchemaFromAst): MetaSchema {
  try {
    const types = schema.types.map((x) => getMetaType(x, schema)).sort((a, b) => a.typename.localeCompare(b.typename))

    return {
      types,
    }
  } catch (ex) {
    throw new GraphQLCosmosInitError(`Error while generating meta schema`, ex)
  }
}

export function getMetaType(type: GraphQLCosmosTypeFromAst, schema: GraphQLCosmosSchemaFromAst): MetaType {
  try {
    return {
      cosmos: type.cosmos,
      external: type.external,
      typename: type.typename,
      database: type.database,
      container: type.container,
      fields: (type.fields ?? [])
        .map((x) => getMetaField(x, type, schema))
        .sort((a, b) => a.fieldname.localeCompare(b.fieldname)),
      filterable: (type.fields ?? []).some((x) => (x.whereOps ?? []).length > 0),
      sortable: (type.fields ?? []).some((x) => x.sortable),
    }
  } catch (ex) {
    throw new GraphQLCosmosInitError(`Type: ${type.typename}`, ex)
  }
}

export function getMetaField(
  field: GraphQLCosmosFieldFromAst,
  type: GraphQLCosmosTypeFromAst,
  schema: GraphQLCosmosSchemaFromAst
): MetaField {
  try {
    const returnType = schema.types.find((x) => x.typename === field.returnTypename) ?? null

    const autoPagination = returnType?.cosmos === true && field.ours === null && field.returnMany
    const pagination = field.pagination === 'on' ? true : field.pagination === 'off' ? false : autoPagination
    return {
      cosmos: field.cosmos,
      returnTypename: field.returnTypename,
      fieldname: field.fieldname,
      database: field.database ?? returnType?.database ?? null,
      pagination: pagination,
      container: field.container ?? returnType?.container ?? null,
      ours: field.ours,
      theirs: field.theirs,
      whereOps: field.whereOps,
      whereOurs: field.whereOurs,
      sortable: field.sortable,
      sortOurs: field.sortOurs,
      returnMany: field.returnMany,
      // kind: toRelationKind(!!field.container, !!field.ours, !!field.theirs, field.returnMany),
    }
  } catch (ex) {
    throw new GraphQLCosmosInitError(`Field: ${field.fieldname}`, ex)
  }
}

// function toRelationKind(container: boolean, ours: boolean, theirs: boolean, list: boolean): MetaFieldKind {
//   if (ours && theirs) {
//     throw Error(`expects ours or theirs not both`)
//   } else if (ours && list) {
//     // if (!container) throw Error(`expects container for many-ours`)
//     return 'many-ours'
//   } else if (ours) {
//     // if (!container) throw Error(`expects container for many-ours`)
//     return 'many-ours'
//   } else if (theirs && list) {
//     // if (!container) throw Error(`expects container for many-theirs`)
//     return 'many-theirs'
//   } else if (ours) {
//     // if (!container) throw Error(`expects container for one-ours`)
//     return 'one-ours'
//   } else if (theirs) {
//     // if (!container) throw Error(`expects container for one-theirs`)
//     return 'one-theirs'
//   } else if (list && container) {
//     return 'many-root'
//   } else {
//     if (container) throw Error(`expects no container for embedded`)
//     if (ours) throw Error(`expects no ours for embedded`)
//     if (theirs) throw Error(`expects no theirs for embedded`)
//     return 'embedded'
//   }
// }
