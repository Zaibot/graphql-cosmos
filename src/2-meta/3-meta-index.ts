import { fail } from '../typescript'
import { MetaField, MetaSchema, MetaType } from './2-intermediate'

export class MetaIndex {
  readonly typeMap: ReadonlyMap<string, MetaType>
  readonly fieldMap: ReadonlyMap<string, MetaField>
  readonly sortFieldMap: ReadonlyMap<string, MetaField>
  readonly whereFieldMap: ReadonlyMap<string, MetaField>
  readonly ourFieldMap: ReadonlyMap<string, MetaField>
  readonly theirFieldMap: ReadonlyMap<string, MetaField>

  constructor(schema: MetaSchema) {
    this.typeMap = new Map(schema.types.map((t) => [t.typename, t]))
    this.fieldMap = new Map(schema.types.flatMap((t) => t.fields.map((f) => [`${t.typename}.${f.fieldname}`, f])))
    this.sortFieldMap = this.buildSortMap(schema)
    this.ourFieldMap = this.buildOurMap(schema)
    this.whereFieldMap = this.buildWhereMap(schema)
    this.theirFieldMap = this.buildTheirMap(schema)
  }

  private buildWhereMap(schema: MetaSchema) {
    const whereFieldMap = new Map()
    for (const type of schema.types) {
      for (const field of type.fields) {
        whereFieldMap.set(`${type.typename}.${field.fieldname}`, field)
        whereFieldMap.set(`${type.typename}.${field.ours ?? field.fieldname}`, field)
        whereFieldMap.set(`${type.typename}.${field.whereOurs ?? field.ours ?? field.fieldname}`, field)
      }
    }
    return whereFieldMap
  }

  private buildOurMap(schema: MetaSchema) {
    const whereFieldMap = new Map()
    for (const type of schema.types) {
      for (const field of type.fields) {
        whereFieldMap.set(`${type.typename}.${field.fieldname}`, field)
        whereFieldMap.set(`${type.typename}.${field.ours ?? field.fieldname}`, field)
      }
    }
    return whereFieldMap
  }

  private buildTheirMap(schema: MetaSchema) {
    const whereFieldMap = new Map()
    for (const type of schema.types) {
      for (const field of type.fields) {
        whereFieldMap.set(`${type.typename}.${field.fieldname}`, field)
        whereFieldMap.set(`${type.typename}.${field.theirs ?? field.fieldname}`, field)
      }
    }
    return whereFieldMap
  }

  private buildSortMap(schema: MetaSchema) {
    const sortFieldMap = new Map<string, MetaField>()
    for (const type of schema.types) {
      for (const field of type.fields) {
        sortFieldMap.set(`${type.typename}.${field.fieldname}`, field)
        sortFieldMap.set(`${type.typename}.${field.ours ?? field.fieldname}`, field)
        sortFieldMap.set(`${type.typename}.${field.sortOurs ?? field.ours ?? field.fieldname}`, field)
      }
    }
    return sortFieldMap
  }

  get allTypes() {
    return Array.from(this.typeMap.values())
  }

  get whereableTypes() {
    return unique(Array.from(this.typeMap.values()).filter((x) => x.filterable))
  }

  get sortableTypes() {
    return unique(Array.from(this.typeMap.values()).filter((x) => x.sortable))
  }

  get pageableTypes() {
    return unique(
      Array.from(this.typeMap.values())
        .flatMap((x) => x.fields)
        .filter((x) => x.pagination)
        .map((x) => this.requireType(x.returnTypename))
    )
  }

  type(typename: string) {
    return this.typeMap.get(typename) ?? null
  }

  field(typename: string, fieldname: string) {
    return this.fieldMap.get(`${typename}.${fieldname}`) ?? null
  }

  whereField(typename: string, fieldname: string) {
    return this.whereFieldMap.get(`${typename}.${fieldname}`) ?? null
  }

  oursField(typename: string, fieldname: string) {
    return this.ourFieldMap.get(`${typename}.${fieldname}`) ?? null
  }

  theirsField(typename: string, fieldname: string) {
    return this.theirFieldMap.get(`${typename}.${fieldname}`) ?? null
  }

  sortField(typename: string, fieldname: string) {
    return this.sortFieldMap.get(`${typename}.${fieldname}`) ?? null
  }

  requireType(typename: string) {
    return this.typeMap.get(typename) ?? fail(`meta type missing: ${typename}`)
  }

  requireField(typename: string, fieldname: string) {
    return this.fieldMap.get(`${typename}.${fieldname}`) ?? fail(`meta field missing: ${typename}.${fieldname}`)
  }
}

function unique<T>(array: T[]) {
  return Array.from(new Set(array).values())
}
