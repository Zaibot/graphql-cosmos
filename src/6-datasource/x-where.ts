import { MetaIndex } from '../2-meta/3-meta-index'
import { fail } from '../typescript'

export type Binary<T = unknown> = [string, T]

export enum WhereOps {
  lt = 'lt',
  ltlower = 'ltlower',
  lte = 'lte',
  ltelower = 'ltelower',
  gt = 'gt',
  gtlower = 'gtlower',
  gte = 'gte',
  gtelower = 'gtelower',
  eq = 'eq',
  eqlower = 'eqlower',
  neq = 'neq',
  neqlower = 'neqlower',
  in = 'in',
  inlower = 'inlower',
  nin = 'nin',
  ninlower = 'ninlower',
  contains = 'contains',
  containslower = 'containslower',
  ncontains = 'ncontains',
  defined = 'defined',
}

export const isWhereOp = (input: string | WhereOps): input is WhereOps => {
  return WhereOpSet.has(input)
}

export interface WhereBinary<T = unknown> {
  lt: Binary<T>
  ltlower: Binary<T>
  lte: Binary<T>
  ltelower: Binary<T>
  gt: Binary<T>
  gtlower: Binary<T>
  gte: Binary<T>
  gtelower: Binary<T>
  eq: Binary<T>
  eqlower: Binary<T>
  neq: Binary<T>
  neqlower: Binary<T>
  in: Binary<T>
  inlower: Binary<T>
  nin: Binary<T>
  ninlower: Binary<T>
  contains: Binary<T>
  containslower: Binary<T>
  ncontains: Binary<T>
  ncontainslower: Binary<T>
  defined: Binary<T>
}

export const WhereOpSet = new Set(Object.values(WhereOps)) as ReadonlySet<string>

export type WhereEntry<T> = Partial<WhereBinary & { and: Where<T>; or: Where<T> }>
export type Where<T = Record<string, unknown>> = Array<WhereEntry<T>>

export function indexWhere(where: Where) {
  let outputId = 0
  let output = new Map<unknown, string>()

  indexWhereRecursive(where)

  function indexWhereRecursive(where: Where) {
    for (const x of where) {
      for (const [op, arg] of Object.entries(x)) {
        if (op === `and`) {
          indexWhereRecursive(arg as Where)
        } else if (op === `or`) {
          indexWhereRecursive(arg as Where)
        } else {
          const [, value] = arg as Binary
          if (output.has(`@p${outputId}`)) {
            output.get(value)
          } else {
            outputId++
            output.set(value, `@p${++outputId}`)
          }
        }
      }
    }
  }

  return output
}
export function transformWhere(
  meta: MetaIndex,
  typename: string,
  output: Map<unknown, string>,
  where: Where,
  alias: string
): string[] {
  return where.flatMap((x) => {
    return Object.entries(x).map(([op, arg]) => {
      if (op === `and`) {
        return `(${transformWhere(meta, typename, output, arg as Where, alias).join(` AND `)})`
      } else if (op === `or`) {
        return `(${transformWhere(meta, typename, output, arg as Where, alias).join(` OR `)})`
      } else {
        const [fieldname, value] = arg as Binary
        const parametername = output.get(value)
        const parametervalue = value
        const field = meta.whereField(typename, fieldname) ?? fail(`requires field ${typename}.${fieldname}`)
        const dbfieldname = field.whereOurs ?? field.ours ?? field.fieldname
        return transformBinary(
          op as keyof WhereBinary,
          [dbfieldname, parametername],
          Array.isArray(parametervalue),
          alias
        )
      }
    })
  })
}

function transformBinary(op: keyof WhereBinary, binary: Binary, secondIsArray: boolean, alias: string) {
  if (op === `defined`) return `IS_DEFINED(${alias}.${binary[0]}) = ${binary[1]}`
  else if (op === `contains`) return `CONTAINS(${alias}.${binary[0]}, ${binary[1]})`
  else if (op === `containslower`) return `CONTAINS(${alias}.${binary[0]}, ${binary[1]}, true)`
  else if (op === `ncontains`) return `CONTAINS(${alias}.${binary[0]}, ${binary[1]})`
  else if (op === `ncontainslower`) return `CONTAINS(${alias}.${binary[0]}, ${binary[1]}, true)`
  else if (op === `eq`) return `${alias}.${binary[0]} = ${binary[1]}`
  else if (op === `eqlower`) return `STRINGEQUALS(${alias}.${binary[0]}, ${binary[1]}, true)`
  else if (op === `lt`) return `${alias}.${binary[0]} < ${binary[1]}`
  else if (op === `ltlower`) return `LOWER(${alias}.${binary[0]}) < LOWER(${binary[1]})`
  else if (op === `lte`) return `${alias}.${binary[0]} <= ${binary[1]}`
  else if (op === `ltelower`) return `LOWER(${alias}.${binary[0]}) <= LOWER(${binary[1]})`
  else if (op === `gt`) return `${alias}.${binary[0]} > ${binary[1]}`
  else if (op === `gtlower`) return `LOWER(${alias}.${binary[0]}) > LOWER(${binary[1]})`
  else if (op === `gte`) return `${alias}.${binary[0]} >= ${binary[1]}`
  else if (op === `gtelower`) return `LOWER(${alias}.${binary[0]}) >= LOWER(${binary[1]})`
  else if (op === `in`)
    return secondIsArray
      ? `ARRAY_CONTAINS(${binary[1]}, ${alias}.${binary[0]})`
      : `ARRAY_CONTAINS(${alias}.${binary[0]}, ${binary[1]})`
  else if (op === `inlower`)
    return secondIsArray
      ? `ARRAY_CONTAINS(${binary[1]}, ${alias}.${binary[0]})`
      : `ARRAY_CONTAINS(${alias}.${binary[0]}, ${binary[1]})`
  else if (op === `neq`) return `${alias}.${binary[0]} != ${binary[1]}`
  else if (op === `neqlower`) return `NOT STRINGEQUALS(${alias}.${binary[0]}, ${binary[1]}, true)`
  else if (op === `nin`)
    return secondIsArray
      ? `NOT ARRAY_CONTAINS(${binary[1]}, ${alias}.${binary[0]})`
      : `NOT ARRAY_CONTAINS(${alias}.${binary[0]}, ${binary[1]})`
  else if (op === `ninlower`)
    return secondIsArray
      ? `NOT ARRAY_CONTAINS(${binary[1]}, ${alias}.${binary[0]})`
      : `NOT ARRAY_CONTAINS(${alias}.${binary[0]}, ${binary[1]})`
  else fail(`unknown op: ${op}`)
}
