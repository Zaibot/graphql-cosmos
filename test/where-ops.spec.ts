import { parse } from 'graphql'
import { MetaIndex } from '../src/2-meta/3-meta-index'
import { indexWhere, transformWhere, WhereBinaryPlural, WhereOps, WhereOpSet } from '../src/6-datasource/x-where'
import { createUnitTestContext } from './utils'

describe(`Where Operation`, () => {
  const ops = Object.values(WhereOps)
  const dummyTypeDefs = parse(`
    type Query {
      dummies: [Dummy] @cosmos
    }

    type Dummy @cosmos(database: "Test", container: "Dummies") {
      id: ID!
      value: String @where(op: "${ops.join(` `)}")
      list: [String] @where(op: "${ops.join(` `)}")
    }
  `)

  const uc = createUnitTestContext(dummyTypeDefs, {})

  for (const op of ops) {
    it(`${op} should match snapshot`, async () => {
      const where = [{ [op]: [`value`, `test`] }]
      const parameters = indexWhere(where)
      const query = transformWhere(new MetaIndex(uc.metaSchema), `Dummy`, parameters, where, `c`).join(` AND `)
      expect({
        query,
        parameters: Object.fromEntries(Array.from(parameters.entries()).map(([x, y]) => [y, x])),
      }).toMatchSnapshot()
    })

    it(`${op} (list result) should match snapshot`, async () => {
      const where = [{ [op]: [`list`, `test`] }]
      const parameters = indexWhere(where)
      const query = transformWhere(new MetaIndex(uc.metaSchema), `Dummy`, parameters, where, `c`).join(` AND `)
      expect({
        query,
        parameters: Object.fromEntries(Array.from(parameters.entries()).map(([x, y]) => [y, x])),
      }).toMatchSnapshot()
    })

    if (WhereBinaryPlural[op]) {
      it(`${op} (plural 1) should match snapshot`, async () => {
        const where = [{ [op]: [`value`, [`test`]] }]
        const parameters = indexWhere(where)
        const query = transformWhere(new MetaIndex(uc.metaSchema), `Dummy`, parameters, where, `c`).join(` AND `)
        expect({
          query,
          parameters: Object.fromEntries(Array.from(parameters.entries()).map(([x, y]) => [y, x])),
        }).toMatchSnapshot()
      })

      it(`${op} (plural 2) should match snapshot`, async () => {
        const where = [{ [op]: [`list`, [`test`]] }]
        const parameters = indexWhere(where)
        const query = transformWhere(new MetaIndex(uc.metaSchema), `Dummy`, parameters, where, `c`).join(` AND `)
        expect({
          query,
          parameters: Object.fromEntries(Array.from(parameters.entries()).map(([x, y]) => [y, x])),
        }).toMatchSnapshot()
      })
    }
  }
})
