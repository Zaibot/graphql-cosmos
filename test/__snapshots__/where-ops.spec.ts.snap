// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Where Operations expects meta schema to remain the same 1`] = `
Object {
  "types": Array [
    Object {
      "container": "Dummies",
      "cosmos": true,
      "database": "Test",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "prop",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
            "in",
          ],
          "whereOurs": null,
        },
      ],
      "filterable": true,
      "sortable": false,
      "typename": "Dummy",
    },
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": "Dummies",
          "cosmos": true,
          "database": "Test",
          "fieldname": "dummies",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Dummy",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Query",
    },
  ],
}
`;

exports[`Where Operations expects schema to remain the same 1`] = `
"schema {
  query: Query
}
type Query {
  dummies(where: DummyWhere, cursor: String, limit: Int): DummyPage!
}
type Dummy {
  id: ID!
  prop: String
}
type DummyPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Dummy]!
}
input DummyWhere {
  and: [DummyWhere!]
  or: [DummyWhere!]
  prop_eq: String
  prop_in: [String]
}"
`;
